dev/cs43l22/export/dev/cs43l22.hpp

```diff
@@ -0,0 +1,838 @@
+        beep_vol_offtime        = 0x1D,
+        beep_tone_cfg           = 0x1E,
+        tone_ctrl               = 0x1F,
+        master_a_vol            = 0x20,
+        master_b_vol            = 0x21,
+        hp_a_vol                = 0x22,
+        hp_b_vol                = 0x23,
+        speak_a_vol             = 0x24,
+        speak_b_vol             = 0x25,
+        ch_mix_swap             = 0x26,
+        limit_ctrl1             = 0x27,
+        limit_ctrl2             = 0x28,
+        limit_attack            = 0x29,
+        ovfl_clk_status         = 0x2E,
+        batt_comp               = 0x2F,
+        vp_batt_level           = 0x30,
+        speak_status            = 0x31,
+        init_sequence_reg2      = 0x32, // used only during init sequence
+        charge_pump_freq        = 0x34,
+        init_sequence_reg3      = 0x47, // used only during init sequence
+    };
+
+    // Writes value into internal codec register
+    static ecl::err register_write(codec_register reg, uint8_t value);
+
+    // Reads internal codec register over I2C
+    static ecl::err register_read(codec_register reg, uint8_t &value);
+
+    // Returns user callback
+    static user_callback &tx_cplt_cb();
+
+    // Internal callback for bus interface
+    static void internal_tx_cplt_cb(ecl::bus_channel ch, ecl::bus_event type, size_t total);
+
+    // State flags.
+    // This bit is set when coded in inited.
+    static constexpr uint8_t inited       = 0x1;
+    // This bit is set when coded is powered.
+    static constexpr uint8_t powered      = 0x2;
+    // This bit is set when there is an active data stream.
+    static constexpr uint8_t tx_active    = 0x4;
+    // This bit is set when user stopped data streams.
+    // Note, that when this bit is set, the tx_active can also be set,
+    // until next TC event will be processed.
+    static constexpr uint8_t tx_stopped   = 0x8;
+    // This bit is set when beep generator is enabled.
+    static constexpr uint8_t beep_enabled = 0x10;
+
+    // Represents the state of the codec.
+    static uint8_t m_state;
+};
+
+template < class I2c, class I2s, class Rst_gpio > uint8_t cs43l22< I2c, I2s, Rst_gpio >::m_state{0};
+
+template < class I2c, class I2s, class Rst_gpio >
+ecl::err cs43l22< I2c, I2s, Rst_gpio >::init()
+{
+    tx_cplt_cb();
+
+    ecl::err rc = I2c::init();
+    if (rc != ecl::err::ok) {
+        return rc;
+    }
+
+    rc = I2s::init();
+
+    if (is_ok(rc)) {
+        m_state |= inited;
+    }
+
+    return rc;
+}
+
+template < class I2c, class I2s, class Rst_gpio >
+ecl::err cs43l22< I2c, I2s, Rst_gpio >::power_up()
+{
+    ecl_assert(m_state & inited);
+
+    Rst_gpio::set();
+
+    uint8_t value {0};
+
+    // power up sequence according to RM
+    ecl::err rc = register_write(pwr_ctrl1, 0x01);
+    //TODO add error checking.
+    register_write(init_sequence_reg1, 0x99);
+    register_write(init_sequence_reg3, 0x80);
+
+    register_read(init_sequence_reg2, value);
```

Comment:
It will be good if you add comment near each register write/read to explain clean purpose of doing these operations (besides that this is an init process in general)